libvast/src/taxonomies.cpp

```diff
@@ -25,163 +25,46 @@
-  if (fs != c->end()) {
-    const auto& fields = caf::get_if<list>(&fs->second);
-    if (!fields)
-      return caf::make_error(ec::convert_error, "fields in", *name,
-                             "is not a list:", fs->second);
-    for (auto& f : *fields) {
-      auto field = caf::get_if<std::string>(&f);
-      if (!field)
-        return caf::make_error(ec::convert_error, "field in", *name,
-                               "is not a string:", f);
-      if (std::count(dest.fields.begin(), dest.fields.end(), *field) > 0)
-        VAST_WARN("ignoring duplicate field for {}",
-                  *name + ": \"" + *field + "\"");
-      else
-        dest.fields.push_back(*field);
-    }
-  }
-  auto cs = c->find("concepts");
-  if (cs != c->end()) {
-    const auto& concepts = caf::get_if<list>(&cs->second);
-    if (!concepts)
-      return caf::make_error(ec::convert_error, "concepts in", *name,
-                             "is not a list:", cs->second);
-    for (auto& c : *concepts) {
-      auto concept_ = caf::get_if<std::string>(&c);
-      if (!concept_)
-        return caf::make_error(ec::convert_error, "concept in", *name,
-                               "is not a string:", c);
-      if (std::count(dest.fields.begin(), dest.fields.end(), *concept_) > 0)
-        VAST_WARN("ignoring duplicate concept for {}",
-                  *name + ": \"" + *concept_ + "\"");
-      else
-        dest.concepts.push_back(*concept_);
-    }
+concept_ mappend(concept_ lhs, concept_ rhs) {
+  if (lhs.description.empty())
+    lhs.description = std::move(rhs.description);
+  else if (!rhs.description.empty() && lhs.description != rhs.description)
+    VAST_WARN("encountered conflicting descriptions: \"{}\" and \"{}\"",
+              lhs.description, rhs.description);
+  for (auto& field : rhs.fields) {
+    if (std::count(lhs.fields.begin(), lhs.fields.end(), field) > 0)
+      VAST_WARN("ignoring duplicate field {}", field);
+    else
+      lhs.fields.push_back(std::move(field));
   }
-  auto desc = c->find("description");
-  if (desc != c->end()) {
-    if (auto description = caf::get_if<std::string>(&desc->second)) {
-      if (dest.description.empty())
-        dest.description = *description;
-      else if (dest.description != *description)
-        VAST_WARN("encountered conflicting descriptions for {}",
-                  *name + ": \"" + dest.description + "\" and \"" + *description
-                    + "\"");
-    }
-  }
-  return caf::none;
-}
-
-caf::error extract_concepts(const data& d, concepts_map& out) {
-  if (const auto& xs = caf::get_if<list>(&d)) {
-    for (const auto& item : *xs) {
-      if (const auto& x = caf::get_if<record>(&item)) {
-        auto n = x->find("concept");
-        if (n == x->end())
-          continue;
-        if (auto err = convert(n->second, out))
-          return err;
-      }
-    }
+  for (auto& c : rhs.concepts) {
+    if (std::count(lhs.concepts.begin(), lhs.concepts.end(), c) > 0)
```

Comment:
```suggestion
    if (lhs.concepts.contains(c))
```

Since C++20, this check is _finally_ idiomatic.

Reply:
Maybe one giant `git grep` for `std::count` would be in place here...

Reply:
Unfortunately this is a `std::vector`. The `contains` member function has only been added to associative containers. We could change the type of this value to `stable_set` and add a `contains` member for that, but I don't want to make this PR even bigger now.